__tests__/integration/utils/test-utils-integration.js

```diff
@@ -806,4 +806,58 @@ export class TestUtils {
 
     return networkHeight;
   }
+
+  /**
+   * Call headless API to get a transaction from the wallet storage
+   *
+   * @param {string} walletId ID of the wallet to get the transaction from
+   * @param {string} txId ID of the transaction to get
+   *
+   * @returns {Promise<*>} The return is an object with { success: boolean }
+   *                       and the transaction object in case of success
+   */
+  static async getTransaction(walletId, txId) {
+    const response = await TestUtils.request
+      .get(`/wallet/transaction/?id=${txId}`)
+      .set(TestUtils.generateHeader(walletId));
+
+    return response.body;
+  }
+
+  /**
+   * Waits until the transaction arrives in the wallet ws, or a timeout (optional)
+   * is reached
+   *
+   * @param {string} walletId ID of the wallet to get the transaction from
+   * @param {string} txId ID of the transaction to get
+   * @param {number | undefined} timeout The timeout to stop waiting for the tx to arrive
+   *                             if not present, it will wait forever
+   *
+   * @returns {Promise<void>} Resolves when the tx is found in the wallet's storage
+   *                          or rejects if timeout is reached
+   */
+  static async waitForTxReceived(walletId, txId, timeout) {
+    /* eslint-disable no-async-promise-executor */
+    return new Promise(async (resolve, reject) => {
+      let timeoutHandler;
+      if (timeout) {
+        // Timeout handler
+        timeoutHandler = setTimeout(() => {
+          reject(new Error(`Timeout of ${timeout}ms without receiving the tx with id ${txId}`));
+        }, timeout);
+      }
+
+      while ((await TestUtils.getTransaction(walletId, txId)).success === false) {
```

Comment:
When a `timeout` is passed as parameter, I believe this `while` block will continue running even after a `reject` is sent. ( [similar case](https://stackoverflow.com/questions/28896280/why-does-javascript-es6-promises-continue-execution-after-a-resolve) on a SO question )

Since this is an util used only on the integration tests, the effects of these "leaks" will probably not be very impactful, but I think it's important to at least document this behavior, so that future improvements can target this easily.

Reply:
From the SO answer I read this:

> Unless an error is thrown, a function is executed until a return statement or its end is reached. Other code outside of the function can't interfere with that (unless, again, an error is thrown).

So if I just add the `return reject` it should be enough, right?

Reply:
Actually, this is not enough because the `return` is ending the `setTimeout` block, I've done some tests here and double-checked it.

I've done the timeout handling a bit differently, what do you think? https://github.com/HathorNetwork/hathor-wallet-headless/pull/308/commits/09783386d74282140e8fea46e2c0e11270a1aa1a

Reply:
I think that's a good solution :+1: